### ISSUE TITLE
Pass clpy_tests/logic_tests/test_truth.py

### ISSUE THREAD

**Issue by ghost** [fixstars/clpy#85]
We should support, remove, or fix clpy_tests/logic_tests/test_truth.py
https://github.com/fixstars/clpy/blob/clpy/tests/clpy_tests/logic_tests/test_truth.py

**Comment by ghost** [comment]
This issue should be investigated after #131 merged (because it provides `clpy.any`) .

**Comment by ghost** [comment]
This test have hangup at 7th item.(in vega)
```
 python -m pytest ./test_truth.py 
=========================================================================================================================== test session starts ============================================================================================================================
platform linux -- Python 3.6.5, pytest-4.2.0, py-1.7.0, pluggy-0.8.1
rootdir: /home/kitawaki/opencl-chainer/clpy, inifile: setup.cfg
collected 432 items                                                                                                                                                                                                                                                        

test_truth.py ......
```

**Comment by LWisteria** [comment]
@t-kitawaki can you provide detail about what the test does?

**Comment by ghost** [comment]
@LWisteria Sorry. I did not read this test carefully. I just only run `python -m pytest ./test_truth.py` to find out this test is easy to solve or not.

**Comment by LWisteria** [comment]
OK. so, @all_developers, please be careful when runing `test_truth.py`.

**Comment by ghost** [comment]
in titanv, this test fail by OpenCLRuntimeError: CL_OUT_OF_RESOURCES
```
$ pytest test_truth.py
============================================================================================= test session starts =============================================================================================
platform linux -- Python 3.6.5, pytest-3.7.3, py-1.6.0, pluggy-0.7.1
rootdir: /home/kitawaki/opencl-chainer/clpy, inifile: setup.cfg
collected 432 items

test_truth.py ......FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF....FFFFFF....FFFFFF....FFFFFF....FFFFFF....FFFFFF....FFFFF [ 42%]
F....FFFFFF....FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF [ 88%]
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF                                                                                                                                                        [100%]

================================================================================================== FAILURES ===================================================================================================
______________________________________________________________________________________ TestAllAny_param_3.test_with_out _______________________________________________________________________________________
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py:500: in test_func
    impl(self, *args, **kw)
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py:124: in test_func
    check_func(clpy_result, numpy_result)
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py:324: in check_func
    array.assert_array_equal(x, y, err_msg, verbose)
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/array.py:90: in assert_array_equal
    clpy.asnumpy(x), clpy.asnumpy(y), err_msg=err_msg,
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/__init__.py:540: in asnumpy
    return a.get(stream=stream)
clpy/core/core.pyx:1677: in clpy.core.core.ndarray.get
    ???
clpy/core/core.pyx:1697: in clpy.core.core.ndarray.get
    ???
clpy/backend/memory.pyx:416: in clpy.backend.memory.MemoryPointer.copy_to_host
    ???
clpy/backend/memory.pyx:417: in clpy.backend.memory.MemoryPointer.copy_to_host
    ???
clpy/backend/opencl/api.pyx:233: in clpy.backend.opencl.api.EnqueueReadBuffer
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   clpy.backend.opencl.exceptions.OpenCLRuntimeError: CL_OUT_OF_RESOURCES

clpy/backend/opencl/exceptions.pyx:23: OpenCLRuntimeError
-------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------
dtype is <class 'numpy.float64'>
_____________________________________________________________________________________ 
...
_________________________________________________________________________________ TestAllAnyWithNaN_param_95.test_without_out _________________________________________________________________________________
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py:500: in test_func
    impl(self, *args, **kw)
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py:103: in test_func
    accept_error=accept_error)
/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py:49: in _check_clpy_numpy_error
    self.fail('Only clpy raises error\n\n' + clpy_tb)
E   AssertionError: Only clpy raises error
E
E   Traceback (most recent call last):
E     File "/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/testing/helper.py", line 24, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/kitawaki/opencl-chainer/clpy/tests/clpy_tests/logic_tests/test_truth.py", line 76, in test_without_out
E       x = xp.asarray(self.x).astype(dtype)
E     File "/home/kitawaki/.pyenv/versions/3.6.5/envs/clpy/lib/python3.6/site-packages/clpy-2.1.0b1-py3.6-linux-x86_64.egg/clpy/creation/from_data.py", line 61, in asarray
E       return core.array(a, dtype, False)
E     File "clpy/core/core.pyx", line 2122, in clpy.core.core.array
E     File "clpy/core/core.pyx", line 2159, in clpy.core.core.array
E     File "clpy/core/core.pyx", line 1730, in clpy.core.core.ndarray.set
E     File "clpy/backend/memory.pyx", line 343, in clpy.backend.memory.MemoryPointer.copy_from_host
E     File "clpy/backend/memory.pyx", line 344, in clpy.backend.memory.MemoryPointer.copy_from_host
E     File "clpy/backend/opencl/api.pyx", line 255, in clpy.backend.opencl.api.EnqueueWriteBuffer
E     File "clpy/backend/opencl/exceptions.pyx", line 23, in clpy.backend.opencl.exceptions.check_status
E   clpy.backend.opencl.exceptions.OpenCLRuntimeError: CL_OUT_OF_RESOURCES
-------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------
dtype is <class 'numpy.float64'>
=================================================================================== 394 failed, 38 passed in 17.42 seconds ====================================================================================
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "clpy/backend/opencl/env.pyx", line 170, in clpy.backend.opencl.env.release
  File "clpy/backend/opencl/api.pyx", line 280, in clpy.backend.opencl.api.Flush
  File "clpy/backend/opencl/exceptions.pyx", line 23, in clpy.backend.opencl.exceptions.check_status
clpy.backend.opencl.exceptions.OpenCLRuntimeError: UNKNOWN ERROR: -9999
```

**Comment by LWisteria** [comment]
I think `CL_OUT_OF_RESOURCES` is correct exception than hang-up (in vega). The problem is, as I said previously https://github.com/fixstars/clpy/issues/85#issuecomment-462624370, what the test is and whether the test really needs huge resources or not (caused by bug).

**Comment by ghost** [comment]
This test only need 2x3x4 ndarray.
https://github.com/fixstars/clpy/blob/f33c5545d72995ffd92dd335dcb112e4e2912bbd/tests/clpy_tests/logic_tests/test_truth.py#L30-L34

So, this test does not need huge resources.
And, CL_OUT_OF_RESOURCES was caused by bug.

**Comment by LWisteria** [comment]
@t-kitawaki Nice work!

**Comment by ghost** [comment]
@LWisteria I'm just only aware of the existence of bug.
I'm still investigating that bug...

**Comment by ghost** [comment]
When I remove 0 size ndarray test (```numpy.zeros((0, 3, 4))```and```numpy.ones((0, 3, 4))```) all tests witch is in TestAllAny is passed (all tests witch is in TestAllAnyWithNaN is not passed).
Passed test count is increase from 44(/432) to 144(/336).
Some test has changed test status (from fail to pass) without any change.
Tests have dependency?

**Comment by ghost** [comment]
Eroor message is different from(better than?) last week ,when I run test in 10 o'clock.
It said that 'any' and 'all' function was worked in clpy and test checking was wrong.
But Error massage is same as last week now(11 o'clock).
It is saying that 'any' and 'all' function is not working...
Test result was changed.

**Comment by ghost** [comment]
Now other reduction functions (lile sum) has same error with 0 size array.
```
Traceback (most recent call last):
  File "/home/kitawaki/opencl-chainer/test/test.py", line 7, in <module>
    print(clpy.sum(x))
  File "clpy/core/core.pyx", line 1658, in clpy.core.core.ndarray.__str__
  File "clpy/core/core.pyx", line 1701, in clpy.core.core.ndarray.get
  File "clpy/backend/memory.pyx", line 416, in clpy.backend.memory.MemoryPointer.copy_to_host
  File "clpy/backend/memory.pyx", line 417, in clpy.backend.memory.MemoryPointer.copy_to_host
  File "clpy/backend/opencl/api.pyx", line 233, in clpy.backend.opencl.api.EnqueueReadBuffer
  File "clpy/backend/opencl/exceptions.pyx", line 23, in clpy.backend.opencl.exceptions.check_status
clpy.backend.opencl.exceptions.OpenCLRuntimeError: CL_OUT_OF_RESOURCES 
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "clpy/backend/opencl/env.pyx", line 170, in clpy.backend.opencl.env.release
  File "clpy/backend/opencl/api.pyx", line 303, in clpy.backend.opencl.api.Flush
  File "clpy/backend/opencl/exceptions.pyx", line 23, in clpy.backend.opencl.exceptions.check_status
clpy.backend.opencl.exceptions.OpenCLRuntimeError: UNKNOWN ERROR: -9999 
```
this function did not raise error at 10 o'clock.

**Comment by ghost** [comment]
solved in #207 
0 size reduction function issue was taken over #208
